### PR TITLE
Support shorthand ambient module declarations

### PR DESCRIPTION
**Pull request by andy-ms**
Half of #6615

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -53,7 +53,8 @@ namespace ts {
             return state;
         }
         else if (node.kind === SyntaxKind.ModuleDeclaration) {
-            return getModuleInstanceState((<ModuleDeclaration>node).body);
+            const body = (<ModuleDeclaration>node).body;
+            return body ? getModuleInstanceState(body) : ModuleInstanceState.NonInstantiated;
```
Comment (mhegazy) [inline src/compiler/binder.ts]: should always be instantiated if there is no body present.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -6295,6 +6295,7 @@ const _super = (function (geti, seti) {
                 write(getGeneratedNameForNode(node));
                 emitEnd(node.name);
                 write(") ");
+                if (!node.body) { throw new Error("TODO"); } //TODO
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: Debug assert.

File: src/compiler/declarationEmitter.ts [hunk 3]
```diff
@@ -834,6 +834,7 @@ namespace ts {
         }
 
         function writeModuleDeclaration(node: ModuleDeclaration) {
+            if (!node.body) { throw new Error("TODO"); } //TODO
```
Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: should be valid. you want to write it the same way you saw it. i.e. short hand declaration.

**Comment by mhegazy** [comment]
please add a test for:
- `import m = require("mod")`
- `export * from "mod"`, where "mod" is shorthand module, then importing from the reexport
- `export  { prop } from "mod"`, where "mod" is shorthand module, then importing from the reexport

**Comment by mhegazy** [comment]
also a test for goto def in the language service.

**Comment by mhegazy** [comment]
other than that looks good. thanks!

**Comment by andy-ms** [comment]
@mhegazy I assume #9077 isn't blocking this?

**Comment by mhegazy** [comment]
> I assume #9077 isn't blocking this?

do not think so.

**Comment by mhegazy** [comment]
👍

**Pull request merged**